packages/server/src/bandit/banditData.ts

```diff
@@ -0,0 +1,133 @@
+        })
+        .promise();
+}
+
+async function getBanditSamplesForTest(testName: string): Promise<TestSample[]> {
+    const queryResult = await queryForTestSamples(testName);
+
+    const parsedResults = queryResultSchema.safeParse(queryResult.Items);
+
+    if (!parsedResults.success) {
+        throw new Error(`Error parsing bandit samples: ${parsedResults.error.toString()}`);
+    }
+
+    return parsedResults.data;
+}
+
+interface BanditVariantData {
+    variantName: string;
+    mean: number;
+}
+
+export interface BanditData {
+    testName: string;
+    bestVariants: BanditVariantData[]; // will contain more than 1 variant if there is a tie
+}
+
+async function buildBanditDataForTest(epicTest: EpicTest): Promise<BanditData> {
+    const samples = await getBanditSamplesForTest(epicTest.name);
+
+    if (samples.length === 0) {
+        // No samples yet, set all means to zero to allow random selection
+        return {
+            testName: epicTest.name,
+            bestVariants: epicTest.variants.map((variant) => ({
+                variantName: variant.name,
+                mean: 0,
+            })),
+        };
+    }
+
+    const variantsData = samples.flatMap((sample) => sample.variants);
+    const variantNames = epicTest.variants.map((variant) => variant.name);
+
+    const variantMeans = variantNames.map((variantName) => {
+        const variantSamples = variantsData.filter(
+            (variantSample) => variantSample.variantName === variantName,
+        );
+
+        const mean = sampleMean(variantSamples);
+
+        return {
+            variantName,
+            mean,
+        };
+    });
+
+    const sortedVariantMeans = variantMeans.sort((a, b) => b.mean - a.mean);
+    const highestMean = sortedVariantMeans[0].mean;
+    const bestVariants = variantMeans.filter((variant) => variant.mean === highestMean);
+
+    return {
+        testName: epicTest.name,
+        bestVariants,
+    };
+}
+
+function sampleMean(samples: VariantSample[]): number {
+    const population = samples.reduce((acc, sample) => acc + sample.views, 0);
+    return samples.reduce(
+        (acc, sample) => acc + (sample.views / population) * sample.avGbpPerView,
+        0,
+    );
+}
+
+function buildBanditData(epicTestsProvider: ValueProvider<EpicTest[]>): Promise<BanditData[]> {
+    const banditTests = epicTestsProvider.get().filter((epicTest) => epicTest.isBanditTest);
+    return Promise.all(
+        banditTests.map((epicTest) =>
+            buildBanditDataForTest(epicTest).catch((error) => {
+                logError(
+                    `Error fetching bandit samples for test ${epicTest.name} from Dynamo: ${error.message}`,
+                );
+                putMetric('bandit-data-load-error');
+                return Promise.reject(error);
+            }),
+        ),
+    );
+}
+
+const buildBanditDataReloader = (epicTestsProvider: ValueProvider<EpicTest[]>) =>
+    buildReloader(() => buildBanditData(epicTestsProvider), 60);
```

Comment:
We could increase the ttl since the sampling only happens hourly. Maybe 5 mins?

Reply:
I don't think it'd make a lot of difference as the source data is only updated hourly on a schedule